docs/design_notes/module_lifecycle.md

```diff
@@ -0,0 +1,289 @@
+
+
+```python
+class Foo(nn.Module):
+
+  @nn.compact
+  def __call__(self, x):
+    dense = nn.Dense(x.shape[-1])
+    return self.inner(x, lambda x: dense(x) + 1)
+  
+  def inner(self, x, fn):
+    for i in range(3):
+      x = fn(x)
+    return x
+```
+
+
+Here `inner` takes a function that closes over a Module instance. In this example, that works fine because we are not transforming the inner method with a lifted transformation. Most methods are not transformed but it is good to know how to make Module methods transformable.
+
+The main obstacle for transformability are types that JAX does not recognize. By default JAX will recognize containers (lists, dictionaries, and tuples) of JAX/NumPy arrays and python scalars (bool, int, float, etc.). The containers are called [pytrees](https://jax.readthedocs.io/en/latest/jax-101/05.1-pytrees.html) and the pytree API is extensible with custom types. With the flax.struct package you can also define pytree compatible dataclasses.
+
+Function closure is the most common way to accidentally hide a JAX array or linen Module from a transformation. There is however an easy workaround if you want to pass closures that are also compatible with JAX and Linen transformations:
+
+
+```python
+class Partial(flax.struct.PytreeNode):
+  fn: Callable = flax.struct.field(pytree_node=False)
+  args: Iterable[Any]
+
+  def __call__(self, *args, **kwargs):
+    return self.fn(*(tuple(self.args) + args), **kwargs)
+
+class Foo(nn.Module):
+
+  @nn.compact
+  def __call__(self, x):
+    dense = nn.Dense(x.shape[-1])
+    return self.inner(x, Partial(lambda mdl, x: mdl(x) + 1, dense)
+  
+  def inner(self, x, fn):
+    ...
+```
+
+
+Here the closure is implemented using a Flax dataclass. The function itself is annotated with `flax.struct.field(pytree_node=False)` to indicate that it does not contain Jax Arrays or linen Modules. The partially applied `args` on the other hand is treated as a pytree container. We rewrite the closure to use Partial. Now the inner method can be transformed using lifted transformations.
+
+
+## Future work
+
+
+### Setup for unbound Modules
+
+The current Module abstraction is particularly restrictive when it comes to initializing fields after construction. In the current Module API, the setup method is the place to initialize the fields of  the Module instance. Because `setup` is only called on a bound Module, the full Module API is available inside setup, including variable declaration. However, oftentimes we don’t actually require any stateful API’s to initialize a field. In fact, most commonly we simply want to declare a submodule. More importantly, it’s often useful to inspect submodules for debugging or to partially run the model. Consider for example:
```

Comment:
Nit: "inside `setup`" <- "inside setup"

```suggestion
The current Module abstraction is particularly restrictive when it comes to initializing fields after construction. In the current Module API, the setup method is the place to initialize the fields of  the Module instance. Because `setup` is only called on a bound Module, the full Module API is available inside `setup`, including variable declaration. However, oftentimes we don’t actually require any stateful API’s to initialize a field. In fact, most commonly we simply want to declare a submodule. More importantly, it’s often useful to inspect submodules for debugging or to partially run the model. Consider for example:
```